keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1273 @@
+
+With the new field `.status.failedIndexes` the Job object can be significantly
+larger as every failed index is recorded in the field. For example, using the
+same format as for the `.status.completedIndexes` and allowing for completions
+equal to `10^6` the size of the field, in the worst-case scenario is 3.4MB, which
+exceeds the etcd limit of 1.5MB.
+
+In order to mitigate this risk we constrain the maximal size of the
+`.status.failedIndexes` and the `.status.completedIndexes` fields by:
+- limiting `.spec.completions` to `10^6`
+- limiting `spec.runPolicy.maxFailedIndexes` to `10^4`; it defaults to `10^4`
+  when completions is greater or equal than `10^4`.
+
+With these limits the worst-case scenario is when:
+- every second index of the last `2*10^5` indexes is in `completedIndexes`
+- every second index of the last `2*10^4` indexes is in `failedIndexes`
+
+In that case, assuming completions is `10^6`, parallelism is `10^5` and
+maxFailedIndexes is `10^4`, the size of the
+`.status.completedIndexes` is limited by 700KB, and `.status.failedIndexes`
+field is limited by 70KB. Together the two fields are limited by 770KB,
+which is less than the etcd limit of `1.5MB`.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+We introduce a new Job API field, called `.spec.runPolicy.backoffLimitPerIndex`.
+When set it limits the number of retries, counted independently for all indexes.
+
+Additionally, we propose the `.spec.runPolicy.maxFailedIndexes` to control
+the maximal number of failed indexes. Once the number is exceeded the entire
+Job is marked Failed and its execution is terminated.
+
+We also propose to extend the PodFailurePolicy with a new action, called
+`FailIndex` to allow an index to fail fast before reaching the backoff limit
+per index.
+
+### Job API
+
+```golang
+
+// PodFailurePolicyAction specifies how a Pod failure is handled.
+// +enum
+type PodFailurePolicyAction string
+
+const (
+  // This is an action which might be taken on a pod failure - mark the
+  // Job's index as failed to avoid restarts within this index. This action
+  // can only be used when backoffLimitPerIndex is set.
+  PodFailurePolicyActionFailIndex PodFailurePolicyAction = "FailIndex"
+  ...
+)
+...
+
+// JobRunPolicy describes custom policy for running a Job
+type JobRunPolicy struct {
+
+  // Specifies the limit for the number of retries within an
+  // index before marking this index as failed. When enabled the number of
+  // failures per index is kept in the pod's
+  // batch.kubernetes.io/job-index-failure-count annotation. It can only
+  // be set to true when Job's completionMode=Indexed, and the Pod's restart
```

Comment:
- Why this is not applied to non-index Job, it means `BackoffLimitPerIndex` works for all kinds of Job? I have no usecases, but still sounds reasonable as independent backOffLimit for independent jobs.
- Why only applied for Pods `restartPolicy=Never`? I guess the answer is  under `#support-backofflimitperindex-when-restartpolicyonfailure`, so still hope to support this feature when restartPolicy=Failure, this is a critical usecase for KEP https://github.com/kubernetes/enhancements/issues/3322

Reply:
> Why this is not applied to non-index Job, it means BackoffLimitPerIndex works for all kinds of Job?

I don't see what would be the meaning of `BackoffLimitPerIndex` for non-Indexed Jobs. I think it would simplify to `backoffLimit`, which we already have.

> Why only applied for Pods restartPolicy=Never?

It is racy if the restarts are controller by the Kubelet, while the limits are controlled by the Job controller (as in the Alternatives section). 

One possibility we could consider is that when `restartPolicy=OnFailure`, and `maxRestartTimes` is specified then we allow to specify at the Job controller level the `maxFailedIndexes` which would control the number of failed indexes (that ran out of restarts by maxRestartTimes)